### ISSUE TITLE
[Bug] Forms.Region.Region.Remove() missing virtual modifier.

### ISSUE THREAD

**Issue by bakerhillpins** [PrismLibrary/Prism#2654]
### Description

I believe the Xamarin.Forms Region class should mimic the WPF Region class' specialization support. I.e. Virtual method surface. To do that the Forms Region.Remove() method needs to be made virtual.

Region.Remove   [Forms ](https://github.com/PrismLibrary/Prism/blob/7f0b1680bbe754da790274f80851265f808d9bbf/src/Forms/Prism.Forms.Regions/Regions/Region.cs#L323)vs [WPF](https://github.com/PrismLibrary/Prism/blob/7f0b1680bbe754da790274f80851265f808d9bbf/src/Wpf/Prism.Wpf/Regions/Region.cs#L277)

I've not completed my attempt yet but it appears that a simpler specialization might be attainable by changing [`Region.GetItemMetadataOrThrow`](https://github.com/PrismLibrary/Prism/blob/7f0b1680bbe754da790274f80851265f808d9bbf/src/Wpf/Prism.Wpf/Regions/Region.cs#L397)  and [`Region.InnerAdd`](https://github.com/PrismLibrary/Prism/blob/7f0b1680bbe754da790274f80851265f808d9bbf/src/Wpf/Prism.Wpf/Regions/Region.cs#L371) to `protected virtual`.  I'll tinker and update if I find anything interesting.

### Steps to Reproduce

See links above to show difference.

### Expected Behavior

Region.Remove is virtual

### Actual Behavior

Region.Remove is not virtual.

### Basic Information

- Version with issue: 8.1.97
- Last known good version: none

**Comment by brianlagunas** [comment]
Why? What is it you are trying to do in which you need to override this method.

Also, I changed this to an enhancement, not a bug.

**Comment by bakerhillpins** [comment]
I'm trying to build out a Region manager that works for a MultiPage<Page>/TabbedPage in Xamarin.  I'm using Prism WPF TabControl as a model of how folks would expect this type of Region to work.  In order to support this I need to track Page objects in conjunction with View objects. This has complexity issues that I'd like to avoid since the Region only ever provides the View object (which could be a Page or a View object on a Multi<Page>) . In the meantime it's helping me understand Regions in general and how to use them.

> Also, I changed this to an enhancement, not a bug.

Help me understand why you consider this be an enhancement? From my perspective I'd think that there would be benefit to keeping Region functionality/support as common as possible between WPF and Forms as possible.  Also, why would that be the only virtual member that's been altered?

**Comment by brianlagunas** [comment]
Your approach is wrong. You don't need a custom region manager. You need a custom region adapter.

A bug is when something is broken or not functioning as expected. This is an enhancement as all you want to do is "enhance" the API to make it more flexible for your specific need. Which probably isn't actually needed in this case.

**Comment by bakerhillpins** [comment]
> Your approach is wrong. You don't need a custom region manager. You need a custom region adapter.

It is not. I have the region adapter created, It's handling all the necessary setup and injects a behavior to handle management of the various properties of the MultiPage<T> class. See the Adapter for TabControl in WPF and associated Selector Behavior as I've basically ported the code from WPF to Xamarin. However, TabControl automatically wraps user objects (views/viewmodels/BOs) in TabItems and then presents the underlying user (bound data) in it's API. Xamarin does not. It's up to the Adapter/Region/Behavior to convert the user view (Could be a Forms.View or a Forms.Page object) into a format that the TabbedPage can support - which is only Forms.Page. That means Wrapping Forms.View objects in a Page and adding them to the control. This mixes wrapped and unwrapped content into the TabbedPage, The Region doesn't allow me to track the wrapper, only the view. Therefore the behavior is contorting to determine which view is being manipulated in the region when views move from Active to Inactive. (collection changed events) This creates a mess of searching and comparisons. I've gone down a few paths and decided I'd see if I could get more information from the Region events. So that's where I'm playing now.

> A bug is when something is broken or not functioning as expected. This is an enhancement as all you want to do is "enhance" the API to make it more flexible for your specific need. Which probably isn't actually needed in this case.

As a consumer of the WPF API it works, As a consumer of the Forms API it doesn't That's a bug. Not a feature. Perspective matters.

**Comment by dansiegel** [comment]
@bakerhillpins your original issue of adding `virtual` is 100% an enhancement. There is nothing broken in Prism. Your issue is one of extensibility which is always going to be an enhancement for us not a bug. Limits imposed on you, by a library you are consuming, are not bugs for the library you're consuming.

**Comment by bakerhillpins** [comment]
@dansiegel We are just going to have to agree that we don't agree on this point.

**Comment by brianlagunas** [comment]
@bakerhillpins both maintainers disagree with you which is what really matter when it comes to this repo 😄 

Also, I am trying to help you by telling you that your approach is wrong. I know what you are trying to do, and I know the issue you are facing with dealing with objects wrapping your actual views and trying to properly add/remove/activate/etc. those views. I've been there and done that. A custom region manager is not what you need. This can all be done with region adapters and behaviors. You can make it harder on yourself it you want, but

If you'd like to submit a PR making this method virtual, we could consider it.  While this request a very easy thing to add, it's not something that will be released to NuGet anytime soon even if added. So keep that in mind as you punish yourself with a custom region manager 😄

**Comment by bakerhillpins** [comment]
@brianlagunas, @dansiegel  Yes, I completely understand that you maintain this code and thus make the final decisions here. You've earned that standing. However, that's not going to keep me from arguing for my perspective or ultimately, that I'm going to like your decision.  😄

> Also, I am trying to help you by telling you that your approach is wrong. I know what you are trying to do, and I know the issue you are facing with dealing with objects wrapping your actual views and trying to properly add/remove/activate/etc. those views. I've been there and done that. A custom region manager is not what you need. This can all be done with region adapters and behaviors.

@brianlagunas So, yes. I DO hear you. I really do. Please understand that I'm simply providing feedback about what I find/observe as I try to learn and solve this problem for myself. I've found very little beyond extremely simple "how to" examples and thus I'm thrashing through several different designs, shifting direction when I hit complexity or bumps. I'm poking the system to see how it reacts, all the while using the WPF source as a guide for functional expectations. The good in all this is it's given me a few more ideas to push on from the Behavior side. So thanks.

In all honesty, I want to have a more meaningful discussion with you both about it but in reality this is burning time you don't have so I'm going to stop now. Frankly, I'd rather have you expend time on a PR that I'll provide and maybe that might help build even a small relationship. 

> If you'd like to submit a PR making this method virtual, we could consider it. While this request a very easy thing to add, it's not something that will be released to NuGet anytime soon even if added. So keep that in mind as you punish yourself with a custom region manager 😄

I plan to. Well, I plan to have a bigger PR with the results of this work and some other stuff with any luck. (after all these years I've grown used to the punishment) Yea, I know that it takes some time to move through the system. But for now this is the available avenue that can result in giving some back.
